Lm eval documentation (#538)
opendatahub-io/opendatahub-documentation: lmeval-evaluation-job.adoc

:_module-type: REFERENCE

ifdef::context[:parent-context: {context}]
[id="lmeval-evaluation-job.adoc_{context}"]
= LM-Eval evaluation job

[role='_abstract']

> LM-Eval service defines a new Custom Resource Definition (CRD) called `LMEvalJob`. An `LMEvalJob` object represents an evaluation job. `LMEvalJob` objects are monitored by the TrustyAI Kubernetes operator.

Reviewer: Where is the LM-Eval service? RHOAI or Openshift? Should it be code formatted?

 Therefore, to run an evaluation job, you first need to create an `LMEvalJob` object with the following information: `model`, `model arguments`, `task`, and `secret`. 

Once the `LMEvalJob` is created, the LM-Eval service will run the evaluation job and update the status and results to the `LMEvalJob` object when the information is available.

[NOTE]
--
It is not recommended to deploy the TrustyAI custom resource (CR) in any namespace that contains non-tabular models. Such models are unsupported by TrustyAI, and may cause errors within the TrustyAI service.
--

.Sample LMEvalJob object 

> Below is an example of an `LMEvalJob` object. 

Reviewer: I wonder if this list might also be better as callouts in the code.

* openai-chat-completions (ChatCompletions API models): Check link:https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/models/openai_completions.py#L209[openai_completions.py] and link:https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/models/api_models.py#L55[tapi_models.py]

textsynth (TextSynth APIs): Check link:https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/models/textsynth.py#L52[textsynth.py]

| `taskList.taskNames`
| Specify a list of tasks supported by `lm-evaluation-harness`.

| `taskList.taskRecipes`
| Specify the task using the Unitxt recipe format:
* `card`: Use the `name` to specify a Unitxt card or `custom` for a custom card
** `name`: Specify a Unitxt card from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the card's ID as the value. For example: The ID of link:https://www.unitxt.ai/en/latest/catalog/catalog.cards.wnli.html[Wnli card] is `cards.wnli`.

> ** `custom`: Define a custom card and use it. The value is a JSON string for a custom Unitxt card which contains the custom dataset. Use the link:https://www.unitxt.ai/en/latest/docs/adding_dataset.html#adding-to-the-catalog[Unitxt documentation] to compose a custom card, store it as a JSON file, and use the JSON content as the value here. If the dataset used by the custom card needs an API key from an environment variable or a persistent volume, you have to set up corresponding resources under the `pod` field. Check the `pod` field below.

Reviewer: ```suggestion
** `custom`: Defines and uses a custom card. The value is a JSON object that contains the custom dataset. For more information about creating a custom card, see the link:https://www.unitxt.ai/en/latest/docs/adding_dataset.html#adding-to-the-catalog[Unitxt documentation]. If the dataset used by the custom card needs an API key from an environment variable or a persistent volume, set up corresponding resources in the `pod` field.
```

> * `template`: Specify a Unitxt template from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the template's ID as the value.
> * `task` (optional): Specify a Unitxt task from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the task's ID as the value. A Unitxt card has a predefined task. Only specify a value for this if you want to run a different task.
> * `metrics` (optional):  Specify a Unitxt task from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the metric's ID as the value. A Unitxt task has a set of pre-defined metrics. Only specify a set of metrics if you need different metrics.
> * `format` (optional): Specify a Unitxt format from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the format's ID as the value.

Reviewer: ```suggestion
* `template`: Specifies a Unitxt template from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the template's ID as the value.
* `task` (optional): Specifies a Unitxt task from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the task's ID as the value. A Unitxt card has a predefined task. Only specify a value for this if you want to run a different task.
* `metrics` (optional):  Specifies a Unitxt task from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the metric's ID as the value. A Unitxt task has a set of pre-defined metrics. Only specify a set of metrics if you need different metrics.
* `format` (optional): Specifies a Unitxt format from the link:https://www.unitxt.ai/en/latest/catalog/catalog.__dir__.html[Unitxt catalog]. Use the format's ID as the value.
```

> * `loaderLimit` (optional): Specifies the maximum number of instances per stream to be returned from the loader (used to reduce loading time in large datasets).

Reviewer: ```suggestion
* `loaderLimit` (optional): Specifies the maximum number of instances per stream to be returned from the loader. You can use this parameter to reduce loading time in large datasets.
```

> * `numDemos` (optional): Number of fewshot to be used.

Reviewer: ```suggestion
* `numDemos` (optional): Number of few-shot to be used.
```

> * `demosPoolSize` (optional): Size of the fewshot pool.

Reviewer: ```suggestion
* `demosPoolSize` (optional): Size of the few-shot pool.
```

| `numFewShot`

> | Sets the number of few-shot examples to place in context. If you are using a task from Unitxt, don't use this field. Use `numDemos` under the `taskRecipes` instead.

Reviewer: ```suggestion
| Sets the number of few-shot examples to place in context. If you are using a task from Unitxt, do not use this field. Use `numDemos` under the `taskRecipes` instead.
```

| `limit`

> | Instead of running the whole dataset, set a limit to run the tasks. Accepts an integer, or a float between 0.0 and 1.0.

Reviewer: ```suggestion
| Set a limit to run the tasks instead of running the entire dataset. Accepts either an integer or a float between 0.0 and 1.0.
```

| `genArgs`

> | Map to `--gen_kwargs` parameter for the lm-evaluation-harness. Read more in the link:https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/interface.md#command-line-interface[lm-evaluation-harness documentation].

Reviewer: ```suggestion
| Maps to the `--gen_kwargs` parameter for the `lm-evaluation-harness`. For more information, see the link:https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/interface.md#command-line-interface[`lm-evaluation-harness` documentation].
```

| `logSamples`

> | If this flag is passed, then the model's outputs, and the text fed into the model, will be saved at per-document granularity.

Reviewer: ```suggestion
| If this flag is passed, then the model's outputs and the text fed into the model will be saved at per-document granularity.
```

| `batchSize`

> | Batch size for the evaluation. The auto:`N` batch size is not used for API models, but numeric batch sizes are used for APIs. Only int batch size supported at the moment.

Reviewer: ```suggestion
| Specifies the batch size for the evaluation in integer format. The `auto:N` batch size is not used for API models, but numeric batch sizes are used for APIs. 
```
Should this be formatted this way? Also guessing at what "int" means here but feel free to correct if I'm off

| `pod`

> | Specify extra information for the lm-eval job's pod. 

Reviewer: ```suggestion
| Specifies extra information for the `lm-eval` job's pod. 
```

Reviewer: Global comment for this doc - present tense instead of imperative and format lm-eval in code

> * `container`: Extra container settings for the lm-eval container.

Reviewer: ```suggestion
* `container`: Specifies additional container settings for the `lm-eval` container.
```

> ** `env`: Specify environment variables. It uses the `EnvVar` data structure of kubernetes.

Reviewer: ```suggestion
** `env`: Specifies environment variables. This parameter uses the `EnvVar` data structure of Kubernetes.
```

> ** `volumeMounts`: Mount the volumes into the lm-eval container

Reviewer: ```suggestion
** `volumeMounts`: Mounts the volumes into the `lm-eval` container
```

** `resources`: Specify the resources for the lm-eval container.

> * `volumes`: Specify the volume information for the lm-eval and other containers. It uses the `Volume`  data structure of kubernetes.

Reviewer: ```suggestion
* `volumes`: Specifies the volume information for the `lm-eval` and other containers. This parameter uses the `Volume` data structure of Kubernetes.
```

> * `sideCars`: A list of containers that run along with the lm-eval container. It uses the `Container` data structure of kubernetes.

Reviewer: ```suggestion
* `sideCars`: A list of containers that run along with the `lm-eval` container. It uses the `Container` data structure of Kubernetes.
```

| `outputs`

> | This section defines custom output locations for the evaluation results storage. At the moment only Persistent Volume Claims (PVC) are supported.

Reviewer: ```suggestion
| This parameter defines a custom output location to store the the evaluation results. Only Persistent Volume Claims (PVC) are supported.
```
Not sure if parameter is the right word

| `outputs.pvcManaged`

> | Create an operator-managed PVC to store this job's results. The PVC will be named `<job-name>-pvc` and will be owned by the `LMEvalJob`. After job completion, the PVC will still be available, but it will be deleted upon deleting the `LMEvalJob`. Supports the following fields:

Reviewer: ```suggestion
| Creates an operator-managed PVC to store this job's results. The PVC is named `<job-name>-pvc` and is owned by the `LMEvalJob`. After the job finishes, the PVC is still be available, but it is deleted with the `LMEvalJob`. Supports the following fields:
```
Can we find a different wording than "is owned by" - in addition to being passive voice, it anthropomorphizes the job. More info: https://www.ibm.com/docs/en/ibm-style?topic=grammar-anthropomorphism
